modules/fedimint-mint-client/src/backup/recovery.rs

```diff
@@ -0,0 +1,757 @@
+        ),
+    >,
+
+    /// Next nonces that we expect might soon get used.
+    /// Once we see them, we move the tracking to `pending_outputs`
+    ///
+    /// Note: since looking up nonces is going to be the most common operation
+    /// the pool is kept shared (so only one lookup is enough), and
+    /// replenishment is done each time a note is consumed.
+    pending_nonces: BTreeMap<CompressedBlindedMessage, (NoteIssuanceRequest, NoteIndex, Amount)>,
+
+    /// Tail of `pending`. `pending_notes` is filled by generating note with
+    /// this index and incrementing it.
+    next_pending_note_idx: Tiered<NoteIndex>,
+
+    /// `LastECashNoteIndex` but tracked in flight. Basically max index of any
+    /// note that got a partial sig from the federation (initialled from the
+    /// backup value). TODO: One could imagine a case where the note was
+    /// issued but not get any partial sigs yet. Very unlikely in real life
+    /// scenario, but worth considering.
+    last_mined_nonce_idx: Tiered<NoteIndex>,
+
+    /// Threshold
+    threshold: u64,
+
+    /// Public key shares for each peer
+    ///
+    /// Used to validate contributed consensus items
+    pub_key_shares: BTreeMap<PeerId, Tiered<PublicKeyShare>>,
+
+    /// Aggregate public key for each amount tier
+    tbs_pks: Tiered<AggregatePublicKey>,
+
+    /// The number of nonces we look-ahead when looking for mints (per each
+    /// amount).
+    gap_limit: u64,
+}
+
+impl MintRestoreInProgressState {
+    fn transitions(
+        &self,
+        _operation_id: OperationId,
+        context: &MintClientContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<MintRestoreStateMachine>> {
+        // TODO: pass refund contexts into state transitions
+        let make_progress_gc = global_context.clone();
+        let secret = context.secret.clone();
+        vec![StateTransition::new(
+            self.clone().make_progress(
+                make_progress_gc.api(),
+                global_context.decoders(),
+                global_context.client_config().epoch_pk,
+                secret,
+            ),
+            move |dbtx, new_state, old_state_machine: MintRestoreStateMachine| {
+                Box::pin(async move {
+                    if new_state.is_done() {
+                        let finalized = new_state.finalize();
+
+                        let dbtx = dbtx.module_tx();
+                        for (amount, note) in finalized.spendable_notes {
+                            let key = NoteKey {
+                                amount,
+                                nonce: note.note.0,
+                            };
+                            dbtx.insert_new_entry(&key, &note).await;
+                        }
+
+                        for (txid, issuance_requests) in finalized.unconfirmed_notes {
+                            // TODO: How do I insert these as new state machines from here?
```

Comment:
@elsirion Please help. Biggest blocker RN.

Reply:
Currently unsupported from inside transitions, we didn't have that use case so far. But since `ModuleGlobalClientContext` has access to the executor via `ClientInner` you should be able to add a method to `IGlobalClientContext` for that.